### PR TITLE
WI #2504 Add partial support for Cobol 6.3 UTF-8

### PR DESCRIPTION
**Pull request by fm-117**
### New features
- `PIC U` syntax is recognized but an errror message is emitted to signal that UTF-8 picture is not supported
  - new `BYTE-LENGTH` phrase is recognized but data is not used
- UTF-8 literals are supported
  - but the max size check is not implemented for non-hexadecimal notation
- usage `UTF-8` and group-usage `UTF-8` are recognized but not supported, specific error message is emitted

### CODE REVIEW

File: TypeCobol.Test/Parser/Programs/Cobol85/UnsupportedUnicodePicture.rdz.cbl [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+       PROGRAM-ID. TCOFM117.
+       DATA DIVISION.
+       WORKING-STORAGE SECTION.
+       01 var1 PIC U USAGE IS UTF-8.
+       01 group1 GROUP-USAGE IS UTF-8.
+          05 part1 PIC U BYTE-LENGTH 20.
+          05 part2 PIC U BYTE-LENGTH 40.
+          05 part3 PIC U BYTE-LENGTH 20.
+       PROCEDURE DIVISION.
+           GOBACK
```
Comment (efr15) [inline TypeCobol.Test/Parser/Programs/Cobol85/UnsupportedUnicodePicture.rdz.cbl:11]: Could you please add here the tests with UTF-8 litterals from `TestTokenTypes.cs`?

**Comment by efr15** [comment]
@Aubertov
How do we manage `PIC U` in **Data Layout** view?
Currently the length displayed in this view is wrong (so the start position of the next data will be wrong as well)

**Review by efr15 — CHANGES_REQUESTED** [review]

File: TypeCobol/AntlrGrammar/CobolCodeElements.g4 [hunk 2]
```diff
@@ -3109,6 +3112,7 @@ usageClause:
 					(DISPLAY_1 NATIVE?) |
 					INDEX |
 					(NATIONAL NATIVE?) |
+					UTF_8 |
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol/AntlrGrammar/CobolCodeElements.g4]: Good thing that one of us reads the specs 🐿️. Done.
Note that for `NATIONAL`, we don't capture the keyword so I did the same for `UTF-8`.

File: TypeCobol/AntlrGrammar/TypeCobolCodeElements.g4 [hunk 3]
```diff
@@ -123,6 +123,7 @@ tcfuncParameterUsageClause:
 					(DISPLAY_1 NATIVE?) |
 					INDEX |
 					(NATIONAL NATIVE?) |
+					UTF_8 |
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol/AntlrGrammar/TypeCobolCodeElements.g4]: yes

File: TypeCobol/Compiler/Scanner/AbstractScanner.cs [hunk 4]
```diff
@@ -361,12 +361,78 @@ protected Token ScanAlphanumericLiteral(int startIndex, TokenType tokenType, Bit
...
+
+        /// <summary>
+        /// Replace escaped Unicode sequences from given character data.
+        /// Escape sequence can be either of the form:
+        /// 
+        /// • \uhhhh, where each h represents a hexadecimal digit in the range '0' to '9', 'a' to 'f', and 'A'
+        /// to 'F', inclusive. This Unicode escape sequence represents a Unicode code point from the Basic
+        /// Multilingual Plane (i.e., Unicode code points in the range U+0000 through U+FFFF).
+        ///
+        /// • \U00hhhhh, where each h represents a hexadecimal digit in the range '0' to '9', 'a' to 'f', and 'A' to 'F'.
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol/Compiler/Scanner/AbstractScanner.cs]: Fixed

File: TypeCobol/Compiler/Types/PictureValidator.cs [hunk 5]
```diff
@@ -363,14 +361,20 @@ private Character[] CollectPictureSequence(List<Tuple<string, int>> matches, Lis
                 }
             }
             bool atLeastOneAGNXZNineStar = symbolCounts[SC.A] + symbolCounts[SC.G] + symbolCounts[SC.N] +
-                                           symbolCounts[SC.X] + symbolCounts[SC.Z] + symbolCounts[SC.NINE] +
-                                           symbolCounts[SC.STAR] >= 1;
+                                           symbolCounts[SC.X] + symbolCounts[SC.U] + symbolCounts[SC.Z] +
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol/Compiler/Types/PictureValidator.cs:364]: done

File: TypeCobol.Test/Parser/Programs/Cobol85/UnsupportedUnicodePicture.rdz.cbl (hunk 1 above)
Comment (fm-117) [reply TypeCobol.Test/Parser/Programs/Cobol85/UnsupportedUnicodePicture.rdz.cbl:17]: Yes. The grammar was lacking the integration of the new literal types.

**Review by efr15 — APPROVED** [review]

**Pull request merged**